hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordMerger.java

```diff
@@ -30,9 +34,19 @@
  * It can implement the merging logic of HoodieRecord of different engines
  * and avoid the performance consumption caused by the serialization/deserialization of Avro payload.
  */
-public interface HoodieMerge extends Serializable {
-  
-  HoodieRecord preCombine(HoodieRecord older, HoodieRecord newer);
+@PublicAPIClass(maturity = ApiMaturityLevel.EVOLVING)
+public interface HoodieRecordMerger extends Serializable {
+
+  /**
+   * This method converges combineAndGetUpdateValue and precombine from HoodiePayload.
+   * It'd be associative operation: f(a, f(b, c)) = f(f(a, b), c) (which we can translate as having 3 versions A, B, C
+   * of the single record, both orders of operations applications have to yield the same result)
+   */
+  Option<HoodieRecord> merge(HoodieRecord older, HoodieRecord newer, Schema schema, Properties props) throws IOException;
 
-  Option<HoodieRecord> combineAndGetUpdateValue(HoodieRecord older, HoodieRecord newer, Schema schema, Properties props) throws IOException;
+  /**
+   * The record type handled by the current merger.
+   * SPARK, AVRO, FLINK
+   */
+  HoodieRecordType getRecordType();
```

Comment:
Imagine I am a Hudi user at Acme Inc. how do I write one merger implementation e.g `AcmeCDCMerger implements HoodieRecordMerger`, and have it be used across engines. IIUC here - we configure one implementation and have it be associated to one record type?

Reply:
Yes, one RecordMerger is associated to one record type.

Reply:
@wzx140 we should actually do it in a similar fashion to how `KeyGenerator` interface is currently implemented:

- You have generic API (`KeyGenerator`) accepting Avro (this is a bare minimum to be implemented, is used as a fallback)
- You have engine specific API (`SparkKeyGeneratorInterface`) which provides for engine-specific APIs that you should implement natively for better performance.

So when Acme implements their own `RecordMerger` they can choose to implement either 

1. A bare-minimum API (Avro), that would allow it to work across the engines but will lack performance
2. Or fully (Avro, Spark, etc) which will be performant across engines

Reply:
Thanks for the suggestion. I think it's great. I've started a discussion in slack to confirm it further.